### ISSUE TITLE
GPU support for computing?

### ISSUE THREAD

**Issue by adeldaoud** [grf-labs/grf#787]
Given that researchers are using datasets of increasing size for causal estimation (see #786), researchers with access to GPU computing would likely be better of using GPU over CPU. I wonder therefore if causal_forest has GPU support? I could not find anything about this in the manual, so I assume the answer is “no, not at the moment.” However, I wonder if there is a workaround enabling me to use my two Nvidia Quadro RTX in parallel? 

And perhaps GPU support may be of interest for a wider audience and a feature to incorporate in some future updates?

**Comment by erikcs** [comment]
Tree growing on a very large data set is more likely to be memory bound than CPU bound: GRF keeps track of many sufficient statistics and stores them in each tree node. This strategy keeps GRF in the realm of data sets ranging to max a few million.

Having a compute agnostic back-end would of course be nice (https://www.jmlr.org/papers/volume21/19-095/19-095.pdf , https://github.com/microsoft/hummingbird), however the goal of the package is valid statistical estimation on moderate data sizes, not `fit(X,y); predict(Xtest)` on 100m observations.

You can speed up GRF by:

1. Bumping up `min.node.size`: grow shallower trees, needs less memory
2. Round/recode the covariates `X` to a lower cardinality: reduces the number of splits to search over

**Comment by adeldaoud** [comment]
Thanks, @erikcs for your input. Here follows a comment and a follow-up question

> Tree growing on a very large data set is more likely to be memory bound than CPU bound: GRF keeps track of many sufficient statistics and stores them in each tree node. This strategy keeps GRF in the realm of data sets ranging to max a few million.
> 
> Having a compute agnostic back-end would of course be nice (https://www.jmlr.org/papers/volume21/19-095/19-095.pdf , https://github.com/microsoft/hummingbird), however the goal of the package is valid statistical estimation on moderate data sizes, not fit(X,y); predict(Xtest) on 100m observations.

I agree with the memory-CPU issue. Based on my experience of GRF, memory has been mainly the issue. However, with data size 3 million rows * 50-ish covariates, CPU limitations start kicking it as well. Some of my GRF models (with more than 4000 trees) take days to run.  While today my data size might be relatively large compared to the data size of the common GRF users, I expect that data size will increase dramatically in the coming year(s) for most users. So more and more users will likely start exploring alternative ways of speeding up their models. As the references you linked to show, using GPU is one good option. Nonetheless, I understand that adding a GPU feature is not a priority at the moment.

> You can speed up GRF by:
> 
> Bumping up min.node.size: grow shallower trees, needs less memory
> Round/recode the covariates X to a lower cardinality: reduces the number of splits to search over

These are good ideas. Would be nice to add a list of such suggestions and examples to the tutorial section of the GRF website. 

A question about tweaking "min.node.size": If I use a GRF with shallower trees, my intuition says that this GRF will likely find less heterogeneity by construction. As a tree will have fewer branches, then that tree will also have fewer interactions among covariates. Fewer interactions imply less (possible) heterogeneity.  Is this intuition correct?

The rounding-recoding suggestion makes sense. Does it also apply to "one-hot encoding" (e.g, using model.matrix()? In other words, I may split a nominal variable X with say five values into five dummy-variables X_1,...X_5 and thus gain variables with lower cardinality than X, but I get a larger covariate matrix. The first matrix has columns of size one, the second has a column of size 5 (four really because we drop one). Thus the GRF speed I gain in lower dimensionality, I will lose with a larger matrix. Does this logic hold?

**Comment by erikcs** [comment]
I don't think there will be a big speed benefit of using many binary variables over one dense one.

**Comment by halflearned** [comment]
Regarding the original request for GPU suppose, we don't any plans of implementing GPU support right now so I'll close this issue. Please re-open as appropriate.
